imap-proto/src/parser/rfc5464.rs

```diff
@@ -0,0 +1,255 @@
+//!
+//! https://tools.ietf.org/html/rfc5464
+//!
+//! IMAP METADATA extension
+//!
+
+#![cfg_attr(rustfmt, rustfmt_skip)]
+#![cfg_attr(feature = "cargo-clippy", allow(redundant_closure))]
+
+
+use nom::IResult;
+
+// mod rfc3501;
+
+use std::str;
+use core::*;
+use types::*;
+use super::rfc3501::*;
+
+fn is_entry_component_char(c: u8) -> bool {
+    c < 0x80 &&
+        c > 0x19 &&
+        c != b'*' &&
+        c != b'%' &&
+        c != b'/'
+}
+
+enum entry_parse_stage<'a>
+{
+    Private_Shared(usize),
+    Admin(usize),
+    Vendor_Comment(usize),
+    Path(usize),
+    Done(usize),
+    Fail(nom::Err<&'a [u8]>),
+}
+
+fn arr_starts_with(a:&[u8], prefix:&[u8]) ->bool
+{
+    let len = prefix.len();
+    if a.len() < len
+    {
+        return false;
+    }
+    for i in 0..len
+    {
+        if a[i] != prefix[i]
+        {
+            return false;
+        }
+    }
+    true
+}
+
+fn check_private_shared(i: &[u8]) -> entry_parse_stage
+{
+    if arr_starts_with(i, b"/private")
+    {
+        return entry_parse_stage::Vendor_Comment(8);
+    }
+    if arr_starts_with(i, b"/shared")
+    {
+        return entry_parse_stage::Admin(7);
+    }
+    return entry_parse_stage::Fail(nom::Err::Error(nom::Context::Code(i, nom::ErrorKind::IsNot)));
+}
+
+fn check_admin(i: &[u8], l:usize) -> entry_parse_stage
+{
+    if arr_starts_with(&i[l..], b"/admin")
+    {
+        return entry_parse_stage::Path(l+6);
+    }
+    return entry_parse_stage::Vendor_Comment(l);
+    
+}
+
+fn check_vendor_comment(i: &[u8], l:usize) ->entry_parse_stage
+{
+    if arr_starts_with(&i[l..], b"/comment")
+    {
+        return entry_parse_stage::Path(l + 8);
+    }
+    if arr_starts_with(&i[l..], b"/vendor")
+    {
+        //make sure vendor name is present
+        if i.len() < l+9 ||
+            i[l+7] != b'/' ||
+            !is_entry_component_char(i[l+8]) {
+                return entry_parse_stage::Fail(nom::Err::Incomplete(nom::Needed::Unknown))
+            }
+        return entry_parse_stage::Path(l+7);
+    }
+    entry_parse_stage::Fail(nom::Err::Error(nom::Context::Code(i, nom::ErrorKind::IsNot)))
+}
+
+
+fn check_path(i: &[u8], l:usize) ->entry_parse_stage
+{
+    if i.len() == l ||
+        i[l] == b' ' ||
+        i[l] == b'\r'
+    {
+        return entry_parse_stage::Done(l);
+    }
+    if i[l] != b'/'
+    {
+        return entry_parse_stage::Fail(nom::Err::Error(nom::Context::Code(i, nom::ErrorKind::IsNot)));
+    }
+    for j in 1..(i.len() - l)
+    {
+        if !is_entry_component_char(i[l+j])
+        {
+            return entry_parse_stage::Path(l+j);
+        }
+    }
+    entry_parse_stage::Done(i.len())
+}
+
+fn check_entry_name(i: &[u8]) -> IResult<&[u8], &[u8]> {
```

Comment:
Do we really need to validate the entire path? The RFC says new paths can be registered with IANA. And that's actually exactly what we want to do with at least the WebPush metadata, maybe even all COI metadata.
Maybe just checking that no `%` or `*` are included is enough?